### PR TITLE
Add generate_stratified_multi_binary_dataset simulator

### LINKED ISSUE #421: Add `generate_stratified_multi_binary_dataset` simulator
# Add `generate_stratified_multi_binary_dataset` simulator

## Background

`glide/simulators/stratified_binary.py`'s `generate_stratified_binary_dataset` generates a batched/stratified dataset for a *single* proxy: it takes `n_samples`/`true_mean`/`proxy_mean`/`correlation` as one value per stratum and loops over `generate_binary_dataset` internally, concatenating the strata together with a `groups` array. `glide/simulators/multi_binary.py`'s `generate_multi_binary_dataset` generates a *single, unstratified* batch for `M >= 1` proxies at once. Nothing generalizes both axes simultaneously: several strata (or batches), each with several proxies.

We need exactly this (a batched production stream monitored by `AsymptoticMultiPPRM`, with two proxies, Gemini and Claude), and mimick hand-rolling the same per-batch loop, `hstack`/`vstack`, and `np.repeat(np.arange(...), ...)` pattern that `generate_stratified_binary_dataset` already implements generically for one proxy. Adding `generate_stratified_multi_binary_dataset` removes that duplication and gives both call sites (and any future multi-proxy monitor tutorial or functional test) a single, tested, reusable generator.

## Design choices

**Location:** `glide/simulators/stratified_multi_binary.py`, alongside `binary.py`/`multi_binary.py`/`stratified_binary.py`.

Mirror `generate_stratified_binary_dataset`'s structure exactly: loop over `K` strata, call the inner per-batch generator once per stratum with `np.random.SeedSequence(random_seed).spawn(K)` for reproducible-but-independent per-stratum seeds, and `hstack`/`vstack` the results together with a `groups` array built via `np.full_like`. The only substitution is the inner generator: `generate_multi_binary_dataset` (per stratum, `M` proxies at once) in place of `generate_binary_dataset` (per stratum, one proxy).

`proxy_means` and `correlations` become 2D, `(K, M)`, one row per stratum — not a single `(M,)` vector shared across all strata. This mirrors `generate_stratified_binary_dataset`'s existing convention that *every* parameter (`n_samples`, `true_mean`, `proxy_mean`, `correlation`) varies per stratum; keeping that consistent is more general (it supports proxy quality drifting over strata too, not just the true mean) than special-casing "same proxies, different `true_mean`." A caller who wants the same proxy profile in every stratum (the common case, including both ticket amendments below) builds it with one `np.tile(proxy_means, (K, 1))` call rather than the function special-casing it.

Validation stays at the strata-shape level only, exactly like `generate_stratified_binary_dataset` does not re-validate what `generate_binary_dataset` already validates per stratum: `_validate_non_empty` on `n_samples`, `_validate_equal_lengths` across `n_samples`/`true_mean`/`proxy_means`/`correlations` (their row counts), and `_validate_is_2d` (`ref-validate-is-2d.md`) on `proxy_means` and `correlations`. Per-stratum feasibility (bounds, correlation feasibility given `true_mean`) is left entirely to the inner `generate_multi_binary_dataset` call, whose errors bubble up unmodified, exactly as `generate_stratified_binary_dataset` already does for `generate_binary_dataset`'s errors. If `ref-validate-is-2d.md` has not merged yet when this ticket is picked up, inline the two-line shape check instead of blocking on it — it's a one-line primitive, not worth stalling this ticket for.

### generate_stratified_multi_binary_dataset

**Location:** `glide/simulators/stratified_multi_binary.py`

**Purpose:** Generate a synthetic batched/stratified binary-label dataset with multiple proxy models, generalizing `generate_stratified_binary_dataset` to `M >= 1` proxies per stratum.

```python
from typing import Optional, Tuple

import numpy as np
from numpy.typing import ArrayLike, NDArray

from glide.core.validation import _validate_equal_lengths, _validate_is_2d, _validate_non_empty
from glide.simulators.multi_binary import generate_multi_binary_dataset


def generate_stratified_multi_binary_dataset(
    n_samples: ArrayLike,
    true_mean: ArrayLike,
    proxy_means: ArrayLike,
    correlations: ArrayLike,
    random_seed: Optional[int] = None,
) -> Tuple[NDArray, NDArray, NDArray]:
    """Generate a synthetic stratified binary-label oracle dataset with multiple proxy models.

    Generalizes ``generate_stratified_binary_dataset`` to ``M >= 1`` proxies per stratum:
    each of the ``K`` strata gets its own ``n_samples``, ``true_mean``, and its own
    length-``M`` row of ``proxy_means``/``correlations``, generated independently by
    ``generate_multi_binary_dataset``, then concatenated together with a stratum identifier.

    Parameters
    ----------
    n_samples : list of int or NDArray of shape (K,)
        Total number of samples per stratum. Length must equal the number of strata.
    true_mean : list of float or NDArray of shape (K,)
        Expected mean value of the true labels per stratum. Length must equal the
        number of strata.
    proxy_means : list of list of float, or NDArray of shape (K, M)
        Expected mean value of each proxy label, per stratum. Row ``k`` is passed as
        ``proxy_means`` to ``generate_multi_binary_dataset`` for stratum ``k``; its
        length ``M`` determines the (fixed, stratum-independent) number of proxies.
    correlations : list of list of float, or NDArray of shape (K, M)
        Pearson correlation between the true label and each proxy label, per stratum.
        Row ``k`` is passed as ``correlations`` to ``generate_multi_binary_dataset``
        for stratum ``k``.
    random_seed : int, optional
        Seed for reproducibility. If provided, one independent child seed is derived
        deterministically per stratum.

    Returns
    -------
    Tuple[NDArray, NDArray, NDArray]
        Let ``N = sum(n_samples)`` be the total number of samples across all strata.

        [0]: array of shape ``(N,)``, y_true containing ground-truth labels.
        [1]: array of shape ``(N, M)``, y_proxies where column m contains proxy labels
             for proxy m across all strata.
        [2]: array of shape ``(N,)``, stratum identifiers.

    Raises
    ------
    ValueError
        - If ``n_samples``, ``true_mean``, ``proxy_means``, and ``correlations`` have
          inconsistent lengths (row counts).
        - If fewer than 1 stratum is specified.
        - If ``proxy_means`` or ``correlations`` is not a 2D array.
        - If any stratum has an infeasible combination of ``true_mean``,
          ``proxy_means``, and ``correlations`` (see ``generate_multi_binary_dataset``).

    Examples
    --------
    >>> import numpy as np
    >>> from glide.simulators import generate_stratified_multi_binary_dataset
    >>> y_true, y_proxies, groups = generate_stratified_multi_binary_dataset(
    ...     n_samples=[6, 8],
    ...     true_mean=[0.6, 0.8],
    ...     proxy_means=[[0.5, 0.55], [0.7, 0.75]],
    ...     correlations=[[0.7, 0.65], [0.75, 0.7]],
    ...     random_seed=42,
    ... )
    >>> len(y_true)
    14
    >>> y_proxies.shape
    (14, 2)
    >>> len(groups)
    14
    """
    n_samples_arr = np.asarray(n_samples, dtype=int)
    true_mean_arr = np.asarray(true_mean, dtype=float)
    proxy_means_arr = np.asarray(proxy_means, dtype=float)
    correlations_arr = np.asarray(correlations, dtype=float)

    _validate_non_empty(n_samples_arr, "n_samples")
    _validate_is_2d(proxy_means_arr, "proxy_means")
    _validate_is_2d(correlations_arr, "correlations")
    num_strata = len(n_samples_arr)

    _validate_equal_lengths(
        n_samples_arr,
        true_mean_arr,
        proxy_means_arr,
        correlations_arr,
        names=["n_samples", "true_mean", "proxy_means", "correlations"],
    )

    y_true_per_stratum = []
    y_proxies_per_stratum = []
    groups_per_stratum = []

    seed_sequence = np.random.SeedSequence(random_seed)
    seeds = seed_sequence.spawn(num_strata)

    for stratum_id in range(num_strata):
        y_true_k, y_proxies_k = generate_multi_binary_dataset(
            n_samples=n_samples_arr[stratum_id],
            true_mean=true_mean_arr[stratum_id],
            proxy_means=proxy_means_arr[stratum_id],
            correlations=correlations_arr[stratum_id],
            random_seed=seeds[stratum_id],
        )
        y_true_per_stratum.append(y_true_k)
        y_proxies_per_stratum.append(y_proxies_k)
        groups_per_stratum.append(np.full_like(y_true_k, stratum_id))

    y_true = np.hstack(y_true_per_stratum)
    y_proxies = np.vstack(y_proxies_per_stratum)
    groups = np.hstack(groups_per_stratum)
    return y_true, y_proxies, groups
```

**Other files:**

| File | What changes |
|---|---|
| `glide/simulators/__init__.py` | Export `generate_stratified_multi_binary_dataset`. |
| `tests/unit/simulators/test_stratified_multi_binary.py` | New, mirroring `tests/unit/simulators/test_stratified_binary.py`: structure/counts, validation delegation, reproducibility, different-seeds-differ. |

## Corner cases

- `proxy_means`/`correlations` must be exactly `(K, M)`; a `(M,)` vector meant to apply the same proxy profile to every stratum must be broadcast by the caller first (`np.tile(proxy_means, (K, 1))`), not passed as-is — this function does not special-case a shared profile.
- The number of proxies `M` is fixed across all strata (`proxy_means.shape[1]`); there is no way to vary the proxy *count* per stratum, matching `MultiPPIMeanEngine`'s own requirement of one consistent `y_proxies` width throughout a stream.
- An infeasible `(true_mean[k], proxy_means[k, m], correlations[k, m])` combination raises from within `generate_multi_binary_dataset` for stratum `k`, unmodified and without indicating which stratum triggered it — mirrors `generate_stratified_binary_dataset`'s existing choice not to wrap or enrich the inner error with stratum context.
- `groups` is returned as a `float` array, not `int`, despite being a stratum identifier — this mirrors `generate_stratified_binary_dataset`'s existing (pre-existing, unrelated to this ticket) behavior via `np.full_like(y_true_k, stratum_id)`; do not diverge from that precedent here.

## Acceptance criteria

- [ ] `generate_stratified_multi_binary_dataset` exists in `glide/simulators/stratified_multi_binary.py` and is exported from `glide.simulators`.
- [ ] Returns `(y_true, y_proxies, groups)` of shapes `(N,)`, `(N, M)`, `(N,)` where `N = sum(n_samples)`, `M = proxy_means.shape[1]`.
- [ ] The same `random_seed` produces identical output across two calls; different seeds produce different output.
- [ ] `doc-asymptotic-multi-pprm-tutorial.md`'s `simulate_batches` helper and `feat-asymptotic-multi-pprm.md`'s functional-test fixture both use this function instead of their manual per-batch loop (see the amendments made to those two tickets).

### PR DESCRIPTION
**Pull request by pravallikamavilla**
### Description

- Adds `generate_stratified_multi_binary_dataset` to `glide/simulators/stratified_multi_binary.py`, generalizing `generate_stratified_binary_dataset` to `M >= 1` proxies per stratum by looping `generate_multi_binary_dataset` once per stratum (reproducible per-stratum seeds via `np.random.SeedSequence(...).spawn(K)`) and concatenating with `hstack`/`vstack` plus a `groups` array, mirroring the existing single-proxy stratified generator exactly.
- Exports the new function from `glide.simulators`, adds it to the API reference index, and updates `CHANGELOG.md`.
- Closes #421

### Checklist

Quality gates that must be satisfied before requesting a review:

- [x] I have read `CONTRIBUTING.md`
- [x] `make lint` passes
- [x] `make type-check` passes
- [x] `make tests` passes
- [x] `make coverage` reports 100% coverage
- [ ] `make test-notebooks` passes
- [x] New public API has numpy-style docstrings
- [x] New public API is inserted in the API reference section of the documentation
- [ ] Docs build without warnings (`make doc`)
- [x] `CHANGELOG.md` updated if the change is user-facing

### LLM usage

Disclose if an LLM was used in writing this PR:
- [ ] No LLM used
- [x] I used an LLM and I went through and validated all the code myself

### CODE REVIEW

File: tests/unit/simulators/test_stratified_multi_binary.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def test_generate_stratified_multi_binary_dataset_different_seed_results_differ():
+    y_true1, y_proxies1, _ = generate_stratified_multi_binary_dataset(
+        n_samples=[5, 5],
+        true_mean=[0.6, 0.8],
+        proxy_means=[[0.5, 0.55], [0.7, 0.75]],
+        correlations=[[0.75, 0.7], [0.75, 0.7]],
+        random_seed=0,
+    )
```
Comment (imerad) [inline tests/unit/simulators/test_stratified_multi_binary.py]: Inputs to `generate_stratified_multi_binary_dataset` are repeated many times in this file. I recommend putting them all into a fixture that returns them as a tuple and passing them with a star expression

File: glide/simulators/stratified_multi_binary.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    n_samples_arr = np.asarray(n_samples, dtype=int)
+    true_mean_arr = np.asarray(true_mean, dtype=float)
+    proxy_means_arr = np.asarray(proxy_means, dtype=float)
+    correlations_arr = np.asarray(correlations, dtype=float)
+
+    _validate_non_empty(n_samples_arr, "n_samples")
+    _validate_is_2d(proxy_means_arr, "proxy_means")
+    _validate_is_2d(correlations_arr, "correlations")
+    num_strata = len(n_samples_arr)
+
```
Comment (imerad) [inline glide/simulators/stratified_multi_binary.py]: For better readability, it would be nice to move the `num_strata` line below where the seed sequence is created.

**Review by imerad — CHANGES_REQUESTED** [review]
Apart from the small remarks, I think the new function should get a functional test of its own. Take the existing ones for multi_binary and stratified_binary as examples.

**Review by imerad — APPROVED** [review]

**Review by gmartinon-ed — APPROVED** [review]
